common/test/matrix/coo_kernels.cpp

```diff
@@ -0,0 +1,475 @@
+        ref_data.expected = gen_mtx(532, num_vectors);
+        ref_data.y = gen_mtx(231, num_vectors);
+        ref_data.complex_b = gen_mtx<ComplexVec>(231, num_vectors);
+        ref_data.complex_x = gen_mtx<ComplexVec>(532, num_vectors);
+        ref_data.alpha = gko::initialize<Vec>({2.0}, ref);
+        ref_data.beta = gko::initialize<Vec>({-1.0}, ref);
+    }
+
+    void unsort_mtx()
+    {
+        gko::test::unsort_matrix(ref_data.mtx.get(), rand_engine);
+    }
+
+    struct data_struct {
+        std::shared_ptr<gko::Executor> exec;
+        std::unique_ptr<Mtx> mtx;
+        std::unique_ptr<Vec> expected;
+        std::unique_ptr<Vec> y;
+        std::unique_ptr<Vec> alpha;
+        std::unique_ptr<Vec> beta;
+        std::unique_ptr<ComplexVec> complex_b;
+        std::unique_ptr<ComplexVec> complex_x;
+
+        data_struct(std::shared_ptr<gko::Executor> exec)
+            : exec(exec),
+              mtx(Mtx::create(exec)),
+              expected(Vec::create(exec)),
+              y(Vec::create(exec)),
+              alpha(Vec::create(exec)),
+              beta(Vec::create(exec)),
+              complex_b(ComplexVec::create(exec)),
+              complex_x(ComplexVec::create(exec))
+        {}
+
+        std::string get_name() const
+        {
+            return gko::name_demangling::get_dynamic_type(*exec);
+        }
+
+        void copy_from(const data_struct &other)
+        {
+            mtx->copy_from(other.mtx.get());
+            expected->copy_from(other.expected.get());
+            y->copy_from(other.y.get());
+            alpha->copy_from(other.alpha.get());
+            beta->copy_from(other.beta.get());
+            complex_b->copy_from(other.complex_b.get());
+            complex_x->copy_from(other.complex_x.get());
+        }
+    };
+
+    template <typename RunFunc, typename TestFunc>
+    void run_test(RunFunc run, TestFunc test)
+    {
+        for (auto &dev : device_data) {
+            SCOPED_TRACE(dev.get_name());
+            dev.copy_from(ref_data);
+        }
+        run(ref_data);
+        for (auto &dev : device_data) {
+            SCOPED_TRACE(dev.get_name());
+            run(dev);
+        }
+        for (auto &dev : device_data) {
+            SCOPED_TRACE(dev.get_name());
+            test(dev, ref_data);
+        }
+    }
+
+    template <typename RunFunc, typename TestFunc>
+    void run_test_returning(RunFunc run, TestFunc test)
+    {
+        for (auto &dev : device_data) {
+            SCOPED_TRACE(dev.get_name());
+            dev.copy_from(ref_data);
+        }
+        auto ref_result = run(ref_data);
+        std::vector<decltype(ref_result)> dev_results;
+        for (auto &dev : device_data) {
+            SCOPED_TRACE(dev.get_name());
+            dev_results.push_back(run(dev));
+        }
+        int i = 0;
+        for (auto &dev : device_data) {
+            SCOPED_TRACE(dev.get_name());
+            test(dev_results[i], ref_result);
+            i++;
```

Comment:
Sometimes, old is gold.
```suggestion
        for (size_t i = 0; i < device_data.size(); i++) {
            SCOPED_TRACE(device_data[i].get_name());
            test(dev_results[i], ref_result);
```